File: parciales/ED1-Parcial1-2021-1--001/codigo/Balanceo.py
from collections import deque
def solve(exp):
  s = deque()
  for i in range(0, len(exp)):
    ei = exp[i]
    if ei == "(" or ei == "{" or ei == "[":
      s.append(ei)
    if ei == ")" or ei == "}" or ei == "]":
      if len(s) == 0 or not valid(s[len(s)-1], ei):
        return False
      else:
        s.pop()
  return len(s) == 0

def valid(o, c):
  return (o == "(" and c == ")") or \
         (o == "[" and c == "]") or \
         (o == "{" and c == "}")

def __main__():
  print(solve("(3 + [a + b])"))
  print(solve("[(6 + {b + c} - 100])"))

__main__()

File: parciales/ED1-Parcial1-2021-1--001/codigo/Balanceo.java
import java.util.LinkedList;
class Balanceo{
public static boolean solve(String exp){
  LinkedList<Character> s = new LinkedList();
  for(int i=0;i<exp.length();++i){
    char ei = exp.charAt(i);
    if(ei == '(' | ei == '{' | ei == '[')
      s.add(ei);
    if(ei == ')' | ei == '}' | ei == ']'){
      if(s.size() == 0 | !valid(s.get(s.size()-1), ei))
        return false;
      else{
        s.remove(s.size()-1);
      }
    }
  }
  return s.size() == 0;
}
public static boolean valid(char o, char c){
  return (o == '(' && c == ')') |
         (o == '[' && c == ']') |
         (o == '{' && c == '}');
}
public static void main(String[] args){
  System.out.println(solve("(3 + [a + b])"));
  System.out.println(solve("[(6 + {b + c} - 100])"));
}
}
